File: project files/server/src/controllers/adminCatalogController.js
const asyncHandler = require('../utils/asyncHandler');
const catalogAdminService = require('../services/catalogAdminService');
const auditLogService = require('../services/auditLogService');

const upsertProduct = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertProduct(req.body, req.auth);
  res.status(200).json({ success: true, data });
});

const deleteProduct = asyncHandler(async (req, res) => {
  await catalogAdminService.deleteProduct(req.params.id);
  res.status(200).json({ success: true, message: 'Product deleted' });
});

const upsertCategory = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertCategory(req.body);
  res.status(200).json({ success: true, data });
});

const deleteCategory = asyncHandler(async (req, res) => {
  await catalogAdminService.deleteCategory(req.params.id);
  res.status(200).json({ success: true, message: 'Category deleted' });
});

const upsertCoupon = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertCoupon(req.body);
  await auditLogService.createAuditLog({
    action: 'COUPON_UPSERT',
    performedBy: req.auth.userId,
    targetType: 'COUPON',
    targetId: data?._id || req.body.id || req.body.code || 'unknown',
    previousValue: null,
    newValue: data,
    req
  });
  res.status(200).json({ success: true, data });
});

const upsertBanner = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertBanner(req.body);
  await auditLogService.createAuditLog({
    action: 'BANNER_UPSERT',
    performedBy: req.auth.userId,
    targetType: 'BANNER',
    targetId: data?._id || req.body.id || req.body.title || 'unknown',
    previousValue: null,
    newValue: data,
    req
  });
  res.status(200).json({ success: true, data });
});

const upsertBrand = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertBrand(req.body);
  res.status(200).json({ success: true, data });
});

const upsertBundleOffer = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertBundleOffer(req.body);
  res.status(200).json({ success: true, data });
});

const upsertSeasonalSale = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertSeasonalSale(req.body);
  res.status(200).json({ success: true, data });
});

const upsertPolicyContent = asyncHandler(async (req, res) => {
  const data = await catalogAdminService.upsertPolicyContent({
    ...req.body,
    updatedBy: req.auth.userId
  });
  res.status(200).json({ success: true, data });
});

module.exports = {
  upsertProduct,
  deleteProduct,
  upsertCategory,
  deleteCategory,
  upsertCoupon,
  upsertBanner,
  upsertBrand,
  upsertBundleOffer,
  upsertSeasonalSale,
  upsertPolicyContent
};

File: spring_back/src/main/java/com/groceryapp/controller/AdminCatalogController.java
package com.groceryapp.controller;

import com.groceryapp.model.*;
import com.groceryapp.security.AuthContext;
import com.groceryapp.service.AuditLogService;
import com.groceryapp.service.CatalogAdminService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/admin")
public class AdminCatalogController {
    private final CatalogAdminService catalogAdminService;
    private final AuditLogService auditLogService;

    public AdminCatalogController(CatalogAdminService catalogAdminService, AuditLogService auditLogService) {
        this.catalogAdminService = catalogAdminService;
        this.auditLogService = auditLogService;
    }

    @PostMapping("/products/upsert")
    public Map<String, Object> upsertProduct(@RequestBody Map<String, Object> body) {
        Product data = catalogAdminService.upsertProduct(body, AuthContext.current().getUserId());
        return Map.of("success", true, "data", data);
    }

    @DeleteMapping("/products/{id}")
    public Map<String, Object> deleteProduct(@PathVariable String id) {
        catalogAdminService.deleteProduct(id);
        return Map.of("success", true, "message", "Product deleted");
    }

    @PostMapping("/categories/upsert")
    public Map<String, Object> upsertCategory(@RequestBody Map<String, Object> body) {
        Category data = catalogAdminService.upsertCategory(body);
        return Map.of("success", true, "data", data);
    }

    @DeleteMapping("/categories/{id}")
    public Map<String, Object> deleteCategory(@PathVariable String id) {
        catalogAdminService.deleteCategory(id);
        return Map.of("success", true, "message", "Category deleted");
    }

    @PostMapping("/coupons/upsert")
    public Map<String, Object> upsertCoupon(@RequestBody Map<String, Object> body, HttpServletRequest request) {
        Coupon data = catalogAdminService.upsertCoupon(body);
        auditLogService.createAuditLog(
                "COUPON_UPSERT",
                AuthContext.current().getUserId(),
                "COUPON",
                data.getId() == null ? str(body.get("code")) : data.getId(),
                null,
                data,
                request
        );
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/banners/upsert")
    public Map<String, Object> upsertBanner(@RequestBody Map<String, Object> body, HttpServletRequest request) {
        Banner data = catalogAdminService.upsertBanner(body);
        auditLogService.createAuditLog(
                "BANNER_UPSERT",
                AuthContext.current().getUserId(),
                "BANNER",
                data.getId() == null ? str(body.get("title")) : data.getId(),
                null,
                data,
                request
        );
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/brands/upsert")
    public Map<String, Object> upsertBrand(@RequestBody Map<String, Object> body) {
        Brand data = catalogAdminService.upsertBrand(body);
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/bundle-offers/upsert")
    public Map<String, Object> upsertBundleOffer(@RequestBody Map<String, Object> body) {
        BundleOffer data = catalogAdminService.upsertBundleOffer(body);
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/seasonal-sales/upsert")
    public Map<String, Object> upsertSeasonalSale(@RequestBody Map<String, Object> body) {
        SeasonalSale data = catalogAdminService.upsertSeasonalSale(body);
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/policies/upsert")
    public Map<String, Object> upsertPolicyContent(@RequestBody Map<String, Object> body) {
        body.put("updatedBy", AuthContext.current().getUserId());
        PolicyContent data = catalogAdminService.upsertPolicyContent(body);
        return Map.of("success", true, "data", data);
    }

    private String str(Object v) {
        return v == null ? "" : String.valueOf(v).trim();
    }
}
